syntax = "proto3";
package eicio.model;
option go_package = "model";

// Warning: do not change EventHeader's fields without understanding how the
// changes affect the EICIO libraries.  Any field may be added without affecting
// the libraries.

message EventHeader {
	uint32 version = 1;

	uint64 runNumber = 2;
	uint64 eventNumber = 3;
	uint64 timeStamp = 4;
	string detector = 5;
	Params params = 6;

	message CollectionHeader {
		string name = 1;
		uint32 id = 2;
		string type = 3;
		uint32 payloadSize = 4;
	}
	repeated CollectionHeader payloadCollections = 7;

	uint32 nUniqueIDs = 8;
	string description = 16;
}

// Message types shared by different collection and entry messages.

message IntParams {
	repeated int32 array = 1;
}

message FloatParams {
	repeated float array = 1;
}

message StringParams {
	repeated string array = 1;
}

message Params {
	map<string, IntParams> ints = 1;
	map<string, FloatParams> floats = 2;
	map<string, StringParams> strings = 3;
}

// References are required by the libraries, and should not be removed.
// However, more fields may be added.
message Reference {
	uint32 collID = 1;
	uint32 entryID = 2;
}

message ParticleID {
	float likelihood = 1;
	int32 type = 2;
	int32 PDG = 3;
	int32 algType = 4;
	repeated float params = 5;
}

// Data model messages.  Each message type must be accompanied by one with the
// identical name suffixed with Collection.  Both message and it's colleciton
// message must have the field "id"

message MCParticle {
	uint32 id = 1;
	repeated Reference parents = 2;
	repeated Reference children = 3;
	int32 PDG = 4;
	repeated double vertex = 5;
	float time = 6;
	repeated double p = 7;
	double mass = 8;
	float charge = 9;
	repeated double PEndPoint = 10;
	repeated float spin = 11;
	repeated int32 colorFlow = 12;

	int32 genStatus = 13;
	uint32 simStatus = 14;
}

message MCParticleCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated MCParticle entries = 4;
}

message SimTrackerHit {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	repeated double pos = 4;
	float EDep = 5;
	float time = 6;
	Reference mc = 7;
	repeated float p = 8;
	float pathLength = 9;
	int32 quality = 10;
}

message SimTrackerHitCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated SimTrackerHit entries = 4;
}

message TrackerRawData {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	int32 time = 4;
	repeated uint32 ADCs = 5;
}

message TrackerRawDataCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated TrackerRawData entries = 4;
}

message TrackerData {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	float time = 4;
	repeated float charges = 5;
}

message TrackerDataCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated TrackerData entries = 4;
}

message TrackerHit {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	int32 type = 4;
	repeated double pos = 5;
	repeated double cov = 6;
	float EDep = 7;
	float EDepErr = 8;
	float time = 9;
	int32 quality = 10;
	repeated Reference rawHits = 11;
}

message TrackerHitCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated TrackerHit entries = 4;
}

message TrackerPulse {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	float time = 4;
	float charge = 5;
	repeated float cov = 6;
	int32 quality = 7;
	Reference TPC = 8;
}

message TrackerPulseCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated TrackerPulse entries = 4;
}

message TrackerHitPlane {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	int32 type = 4;
	repeated double pos = 5;
	repeated float U = 6;
	repeated float V = 7;
	float dU = 8;
	float dV = 9;
	float EDep = 10;
	float EDepErr = 11;
	float time = 12;
	int32 quality = 13;
	repeated Reference rawHits = 14;
}

message TrackerHitPlaneCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated TrackerHitPlane entries = 4;
}

message TrackerHitZCylinder {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	int32 type = 4;
	repeated double pos = 5;
	repeated float center = 6;
	float dRPhi = 7;
	float dZ = 8;
	float EDep = 9;
	float EDepErr = 10;
	float time = 11;
	int32 quality = 12;
	repeated Reference rawHits = 13;
}

message TrackerHitZCylinderCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated TrackerHitZCylinder entries = 4;
}

message Track {
	uint32 id = 1;
	int32 type = 2;
	float chi2 = 3;
	int32 NDF = 4;
	float dEdx = 5;
	float dEdxErr = 6;
	float radius = 7;
	repeated int32 subDetHits = 8;

	message TrackState {
		int32 loc = 1;
		float d0 = 2;
		float phi = 3;
		float omega = 4;
		float z0 = 5;
		float tanL = 6;
		repeated float cov = 7;
		repeated float ref = 8;
	}
	repeated TrackState states = 9;

	repeated Reference hits = 10;
	repeated Reference tracks = 11;
}

message TrackCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated Track entries = 4;
}

message SimCalorimeterHit {
	uint32 id = 1;
	Params params = 2;
	int32 cellID0 = 3;
	int32 cellID1 = 4;
	float energy = 5;
	repeated float pos = 6;

	message Contrib {
		Reference MCParticle = 1;
		float energy = 2;
		float time = 3;
		int32 PDG = 4;
		repeated float stepPos = 5;
	}
	repeated Contrib contributions = 7;
}

message SimCalorimeterHitCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated SimCalorimeterHit entries = 4;
}

message RawCalorimeterHit {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	int32 amplitude = 4;
	int32 timeStamp = 5;
}

message RawCalorimeterHitCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated RawCalorimeterHit entries = 4;
}

message CalorimeterHit {
	uint32 id = 1;
	int32 cellID0 = 2;
	int32 cellID1 = 3;
	float energy = 4;
	float energyErr = 5;
	float time = 6;
	repeated float pos = 7;
	int32 type = 8;
	Reference raw = 9;
}

message CalorimeterHitCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated CalorimeterHit entries = 4;
}

message Cluster {
	uint32 id = 1;
	int32 type = 2;
	float energy = 3;
	float energyErr = 4;
	repeated float pos = 5;
	repeated float posErr = 6;
	float theta = 7;
	float phi = 8;
	repeated float dirErr = 9;
	repeated float shape = 10;
	repeated ParticleID PIDs = 11;
	repeated Reference clusters = 12;
	repeated Reference hits = 13;
	repeated float weights = 14;
	repeated float subDetEnes = 15;
}

message ClusterCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated Cluster entries = 4;
}

message RecParticle {
	uint32 id = 1;
	int32 type = 2;
	repeated float p = 3;
	float energy = 4;
	repeated float cov = 5;
	float mass = 6;
	float charge = 7;
	repeated float ref = 8;
	repeated ParticleID PIDs = 9;
	int32 PIDUsed = 10;
	float goodnessOfPID = 11;
	repeated Reference recs = 12;
	repeated Reference tracks = 13;
	repeated Reference clusters = 14;
	Reference startVtx = 15;
}

message RecParticleCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated RecParticle entries = 4;
}

message Vertex {
	uint32 id = 1;
	int32 primary = 2;
	int32 algType = 3;
	float chi2 = 4;
	float prob = 5;
	repeated float pos = 6;
	repeated float cov = 7;
	repeated float params = 8;
	Reference recPart = 9;
}

message VertexCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated Vertex entries = 4;
}

message Relation {
	uint32 id = 1;
	Reference from = 2;
	Reference to = 3;
	float weight = 4;
}

message RelationCollection {
	uint32 id = 1;
	uint32 flags = 2;
	Params params = 3;
	repeated Relation entries = 4;
}
